pkg/jobrunaggregator/releasebigqueryloader/release_uploader.go

```diff
@@ -0,0 +1,225 @@
+package releasebigqueryloader
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+
+	"cloud.google.com/go/bigquery"
+	"cloud.google.com/go/civil"
+	"github.com/pkg/errors"
+
+	"github.com/openshift/ci-tools/pkg/jobrunaggregator/jobrunaggregatorlib"
+)
+
+type allReleaseUploaderOptions struct {
+	ciDataClient jobrunaggregatorlib.CIDataClient
+	// GCSClient is used to read the prowjob data
+	gcsClient  jobrunaggregatorlib.CIGCSClient
+	httpClient *http.Client
+
+	releaseInserter *bigquery.Inserter
+	releaseTable    *bigquery.Table
+
+	releaseJobRunInserter *bigquery.Inserter
+	releaseJobRunTable    *bigquery.Table
+
+	releases  []string
+	ciDataSet *bigquery.Dataset
+}
+
+func (r *allReleaseUploaderOptions) Run(ctx context.Context) error {
+	if err := r.findOrCreateTable(ctx, jobrunaggregatorlib.ReleaseTableName); err != nil {
+		return errors.Wrapf(err, "could not find or create %s table", jobrunaggregatorlib.ReleaseTableName)
+	}
+
+	if err := r.findOrCreateTable(ctx, jobrunaggregatorlib.ReleaseJobRunTableName); err != nil {
+		return errors.Wrapf(err, "could not find or create %s table", jobrunaggregatorlib.ReleaseJobRunTableName)
+	}
+
+	releaseTagSet, err := r.ciDataClient.ListReleaseTags(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, release := range r.releases {
+		fmt.Fprintf(os.Stderr, "Fetching release %s from release controller...\n", release)
+		tags := r.fetchReleaseTags(release)
+
+		for _, tag := range tags.Tags {
+			fmt.Fprintf(os.Stderr, "Fetching tag %s from release controller...\n", tag.Name)
+			releaseDetails := r.fetchReleaseDetails(release, tag)
+			bigQueryReleaseDetails := releaseDetailsToBigQuery(tag, releaseDetails)
+
+			// We skip releases that aren't fully baked, or already in the big query tables:
+			if _, ok := releaseTagSet[bigQueryReleaseDetails.ReleaseTag]; ok {
+				continue
+			}
+
+			if bigQueryReleaseDetails.Phase == "Ready" || !bigQueryReleaseDetails.ChangeLog.Valid {
+				continue
+			}
+
+			if err := r.releaseInserter.Put(ctx, bigQueryReleaseDetails); err != nil {
+				return errors.Wrapf(err, "could not insert release details for %s", tag.Name)
+			}
+			for _, run := range releaseJobRunsToBigQuery(releaseDetails) {
```

Comment:
how do you know all the jobs have completed?

Reply:
If a release is anything but Ready, it's done -- confirmed on Slack with CRT folks.  We also check to make sure the changelog is non-null, since it can take some time to generate.